src/main/kotlin/io/libp2p/core/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName, NoiseIoHandshake())
+        ch.nettyChannel.pipeline().addLast(handshakeHandlerName + "ResultHandler", resultHandler)
+        return ret
+    }
+
+    inner class NoiseIoHandshake : SimpleChannelInboundHandler<ByteBuf>() {
+        private val handshakestate: HandshakeState = HandshakeState(protocolName, role)
+
+        init {
+            handshakestate.localKeyPair.copyFrom(localDHState)
+            handshakestate.start()
+            logger.debug("Starting handshake")
+        }
+
+        override fun channelRead0(ctx: ChannelHandlerContext?, msg: ByteBuf?) {
+            channelRead(ctx!!, msg as Any)
+            super.channelRead(ctx, msg)
+        }
+
+        override fun channelRegistered(ctx: ChannelHandlerContext?) {
+            super.channelRegistered(ctx)
+
+            if (activated) {
+                return
+            }
+            logger.debug("Registration starting")
+            activated = true
+
+            if (role == HandshakeState.INITIATOR) {
+                val msgBuffer = ByteArray(65535)
+
+                // TODO : include data fields into protobuf struct to match spec
+                // alice needs to put signed peer id public key into message
+                val signed = localKey.sign(localKey.publicKey().bytes())
+
+                // generate an appropriate protobuf element
+                val bs = Spipe.Exchange.newBuilder().setEpubkey(ByteString.copyFrom(localKey.publicKey().bytes()))
+                    .setSignature(ByteString.copyFrom(signed)).build()
+
+                // create the message
+                // also create assign the signed payload
+                val msgLength = handshakestate.writeMessage(msgBuffer, 0, bs.toByteArray(), 0, bs.toByteArray().size)
+
+                // put the message frame which also contains the payload onto the wire
+                val writeAndFlush = ctx?.writeAndFlush(msgBuffer.copyOfRange(0, msgLength))
+                writeAndFlush?.await()
+            }
+            logger.debug("Registration complete")
+        }
+
+        override fun channelActive(ctx: ChannelHandlerContext?) {
+            logger.debug("Activation starting")
+            super.channelActive(ctx)
+            channelRegistered(ctx)
+            logger.debug("Activation complete")
+        }
+
+        private var activated = false
+        private var flagRemoteVerified = false
+        private var flagRemoteVerifiedPassed = false
+        private var aliceSplit: CipherState? = null
+        private var bobSplit: CipherState? = null
+        private var cipherStatePair: CipherStatePair? = null
+
+        override fun channelRead(ctx: ChannelHandlerContext, msg1: Any) {
+            logger.debug("Starting channelRead0")
+            val msg = if (msg1 is ByteArray) {
+                msg1
+            } else {
+                (msg1 as ByteBuf).array()
+            }
```

Comment:
Only `ByteBuf` messages are expected in this handler.

Reply:
done